### PR TITLE
Allow tracing-serde to work on no_std.

### PR DESCRIPTION
**Pull request by Hoverbear**
Allows `tracing-serde` to work on devices without `std`.

*  Enable `#[no_std]` in `tracing-serde` when `cfg(not(feature = std))`
*  No longer have unnecessary dependencies on std in serde/tracing.

No behavior differences or actual code changes other than allowing for core/alloc fallbacks.

### CODE REVIEW

**Comment by hawkw** [comment]
I was concerned that this might be a breaking change for anyone who depended on `tracing-serde` with `default-features = false`, but I realize we didn't previously have any feature flags for that crate...I _think_ this is not breaking. 

Do you happen to know if it's allowed to set `default-features = false` for a crate with no feature flags? If it is, anyone who has done that would no longer have `serde`'s std feature enabled, unless enabled elsewhere...

**Comment by Hoverbear** [comment]
It seems to be allowed.

image

**Comment by hawkw** [comment]
> It seems to be allowed.
> 
> image

Hmm. Is it possible that this could break anything? I'm not sure. 

_Presumably_, if you're using `tracing-serde`, you're also depending on `serde` (or `serde-json` or something), and those dependencies would enable `serde`'s default features unless you've explicitly disabled them, so disabling the default features on `tracing-serde`'s `serde` dependency wouldn't break APIs you expect to be there...right? 

Sorry for drilling so deep into this, I'd just really like to work out whether we can ship this change in a patch or if it has to be a breaking release. Normally we'd definitely consider feature flagging something that wasn't previously feature flagged a breaking change, but in this case, I'm not sure...

**Comment by Hoverbear** [comment]
I absolutely think it's fine to wait for a breaking compatible release for this. I don't want to break anyone's production code just so I don't need a Git dependency for a few weeks. :)

Thanks so much Eliza.

File: tracing-serde/src/lib.rs
Comment (jyn514) [inline tracing-serde/src/lib.rs:34]: nit: `core` works in both, no need to duplicate the code. `std` is mostly a facade crate.
```suggestion
use core::fmt;
```

**Comment by hawkw** [comment]
Once #998 merge, everything on the master branch will be versioned for 0.2.x, so we can actually land this soon for `tracing-serde` 0.2

File: tracing-serde/src/lib.rs
Comment (hawkw) [reply tracing-serde/src/lib.rs:34]: we should probably do this in some of the other crates, which currently have their own weird facade modules to import names from either `std` or `core` conditionally. those are probably not necessary?

File: tracing-serde/Cargo.toml
Comment (hawkw) [inline tracing-serde/Cargo.toml:22]: Can we add a list of feature flags in the crate's `lib.rs` docs and mention the existence of this one?

**Review by hawkw — COMMENTED** [review]
Mind rebasing this onto the latest master and adding the feature flag to the docs? Once that's taken care of, this should be good to merge!

**Comment by Hoverbear** [comment]
Thanks for the feedback! Comments resolved. :)

**Review by hawkw — APPROVED** [review]
lovely, thanks!

**Pull request merged**
